### ISSUE TITLE
unused-argument should report overwritten values

### ISSUE THREAD

**Issue by matejcik** [PyCQA/pylint#3102]
The following code gets 10/10 score, but is trivially incorrect:
```python
def hello(who):
    """Override argument without using it"""
    who = "world"
    print(who)

def foobar():
    """Override variable value without using it"""
    quux = "a"
    quux = "b"
    return quux
```

The first instance bit me today, a function was taking an argument but overwriting its value immediately. IMHO it should either report `unused-argument` or some other warning.

Not sure about the second instance, but including it for completeness.

I'm not sure if detecting this is possible; consider the following code:
```python
def correct_func(flag, arg):
	if flag:
        arg = "overwritten"
    return arg

def incorrect_func(flag, arg):
    if flag:
        arg = "overwritten if true"
    else:
        arg = "also overwritten if false"
    return arg
```
is pylint even capable of determining that on at least one path, the original argument value is used unchanged?

**Comment by PCManticore** [comment]
Thanks opening an issue!

The first one makes sense and it's trivial to emit an `unused-argument` there. The latter one might need some basic support for flow control understanding to get it right in most cases.

**Comment by clavedeluna** [comment]
Started to look at this issue, at least to be able to tackle the very first test case which seemed trivial, but it isn't quite so (at least for me right now). Leaving some more test cases here for the future.

```
def hello(who): # should be [unused-argument]
    """Override argument without using it"""
    who = "world"
    print(who)


def hello(who): # should be [unused-argument]
    """Override argument without using it"""
    # some space some comments
    print("someone")
    who = "world"
    print(who)

def hello(who="mundo"): # should be [unused-argument]
    """Override argument without using it"""
    who = "world"
    print(who)

def hello(who, msg): # should be [unused-argument, unused-argument]
    """Override argument without using it"""
    who = "world"
    msg = "hello" # should be [unused-variable]
    print(who)


def hello(who, msg):
    print(who)
    print(msg)
    who = "world"
    msg = "hello"
    return who, msg
```
